File: src/harness_core/dispatch/routing/schemas.py
"""Phase 4 routing schemas: experiments, arms, observations, promotion verdicts."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

# ---------------------------------------------------------------------------
# Schema versions
# ---------------------------------------------------------------------------

ROUTING_EXPERIMENT_SCHEMA_VERSION = "routing_experiment.v1"
ROUTING_ARM_SCHEMA_VERSION = "routing_arm.v1"
ROUTING_OBSERVATION_SCHEMA_VERSION = "routing_observation.v1"
PROMOTION_VERDICT_SCHEMA_VERSION = "promotion_verdict.v1"

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

EXPERIMENT_STATUSES: tuple[str, ...] = ("created", "running", "concluded", "rolled_back")
EXPERIMENT_CONCLUSIONS: tuple[str, ...] = ("adopt_candidate", "keep_baseline", "inconclusive", "rolled_back")
ROUTING_MODES: tuple[str, ...] = ("static", "adaptive", "shadow")
PROMOTION_VERDICTS: tuple[str, ...] = ("promote", "hold", "reject", "insufficient_data")
DOWNGRADE_REASONS: tuple[str, ...] = ("cost_optimization", "quality_sufficient", "budget_pressure")
UPGRADE_REASONS: tuple[str, ...] = ("high_uncertainty", "failure_rate", "critical_task", "quality_risk")

PROMOTION_GATE_DEFAULTS: dict[str, Any] = {
    "min_sample_count": 30,
    "max_failure_rate_delta": 0.05,
    "min_cost_reduction_pct": 5.0,
}

# ---------------------------------------------------------------------------
# Task group helpers — "/" delimiter to avoid collision with underscored domains
# ---------------------------------------------------------------------------

_TASK_GROUP_SEP = "/"


def make_task_group(domain: str, intent: str) -> str:
    return f"{domain}{_TASK_GROUP_SEP}{intent}"


def parse_task_group(task_group: str) -> tuple[str, str]:
    parts = task_group.split(_TASK_GROUP_SEP, 1)
    domain = parts[0] if parts else ""
    intent = parts[1] if len(parts) > 1 else ""
    return domain, intent

# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class RoutingObservation:
    observation_id: str
    arm_id: str
    dispatch_id: str
    task_domain: str
    task_intent: str
    selected_tier: str
    baseline_tier: str
    quality_score: float
    cost: float
    latency_ms: int
    success: bool
    failure_domain: str | None = None
    budget_violation: bool = False
    observed_at: str = ""
    schema_version: str = ROUTING_OBSERVATION_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "observation_id": self.observation_id,
            "arm_id": self.arm_id,
            "dispatch_id": self.dispatch_id,
            "task_domain": self.task_domain,
            "task_intent": self.task_intent,
            "selected_tier": self.selected_tier,
            "baseline_tier": self.baseline_tier,
            "quality_score": self.quality_score,
            "cost": self.cost,
            "latency_ms": self.latency_ms,
            "success": self.success,
            "failure_domain": self.failure_domain,
            "budget_violation": self.budget_violation,
            "observed_at": self.observed_at,
        }


@dataclass(frozen=True)
class RoutingArm:
    arm_id: str
    experiment_id: str
    tier: str
    profile_id: str | None = None
    traffic_weight: float = 1.0
    observations: tuple[RoutingObservation, ...] = ()
    schema_version: str = ROUTING_ARM_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "arm_id": self.arm_id,
            "experiment_id": self.experiment_id,
            "tier": self.tier,
            "profile_id": self.profile_id,
            "traffic_weight": self.traffic_weight,
            "observations": [o.to_dict() for o in self.observations],
        }


@dataclass(frozen=True)
class RoutingExperiment:
    experiment_id: str
    name: str
    task_group: str
    arms: tuple[RoutingArm, ...] = ()
    status: str = "created"
    start_time: str | None = None
    end_time: str | None = None
    conclusion: str | None = None
    schema_version: str = ROUTING_EXPERIMENT_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "experiment_id": self.experiment_id,
            "name": self.name,
            "task_group": self.task_group,
            "arms": [a.to_dict() for a in self.arms],
            "status": self.status,
            "start_time": self.start_time,
            "end_time": self.end_time,
            "conclusion": self.conclusion,
        }


@dataclass(frozen=True)
class PromotionVerdict:
    verdict: str
    task_group: str
    candidate_tier: str
    baseline_tier: str
    sample_count: int
    quality_delta: float
    cost_reduction_pct: float
    failure_rate_delta: float
    reasons: tuple[str, ...] = ()
    requires_human_review: bool = False
    schema_version: str = PROMOTION_VERDICT_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "verdict": self.verdict,
            "task_group": self.task_group,
            "candidate_tier": self.candidate_tier,
            "baseline_tier": self.baseline_tier,
            "sample_count": self.sample_count,
            "quality_delta": self.quality_delta,
            "cost_reduction_pct": self.cost_reduction_pct,
            "failure_rate_delta": self.failure_rate_delta,
            "reasons": list(self.reasons),
            "requires_human_review": self.requires_human_review,
        }


@dataclass
class RoutingSelection:
    """Return type from adaptive tier selectors — carries routing metadata."""

    selected_tier: str
    selected_profile_id: str | None
    fallback_tier: str
    fallback_profile_id: str | None
    shadow_routes: list  # list[ShadowRoute]
    rejected_candidates: list  # list[RejectedCandidate]
    routing_reason: str
    routing_mode: str  # "adaptive" or "static"
    routing_experiment_id: str | None = None

    def as_tuple_7(self) -> tuple:
        return (
            self.selected_tier,
            self.selected_profile_id,
            self.fallback_tier,
            self.fallback_profile_id,
            self.shadow_routes,
            self.rejected_candidates,
            self.routing_reason,
        )

File: engine/src/routing/schemas.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

// Schema versions
pub const ROUTING_EXPERIMENT_SCHEMA_VERSION: &str = "routing_experiment.v1";
pub const ROUTING_ARM_SCHEMA_VERSION: &str = "routing_arm.v1";
pub const ROUTING_OBSERVATION_SCHEMA_VERSION: &str = "routing_observation.v1";
pub const PROMOTION_VERDICT_SCHEMA_VERSION: &str = "promotion_verdict.v1";

// Constants
pub const EXPERIMENT_STATUSES: &[&str] = &["created", "running", "concluded", "rolled_back"];
pub const EXPERIMENT_CONCLUSIONS: &[&str] = &[
    "adopt_candidate",
    "keep_baseline",
    "inconclusive",
    "rolled_back",
];
pub const ROUTING_MODES: &[&str] = &["static", "adaptive", "shadow"];
pub const PROMOTION_VERDICTS: &[&str] = &["promote", "hold", "reject", "insufficient_data"];
pub const DOWNGRADE_REASONS: &[&str] =
    &["cost_optimization", "quality_sufficient", "budget_pressure"];
pub const UPGRADE_REASONS: &[&str] = &[
    "high_uncertainty",
    "failure_rate",
    "critical_task",
    "quality_risk",
];

pub const PROMOTION_GATE_MIN_SAMPLE_COUNT: usize = 30;
pub const PROMOTION_GATE_MAX_FAILURE_RATE_DELTA: f64 = 0.05;
pub const PROMOTION_GATE_MIN_COST_REDUCTION_PCT: f64 = 5.0;

const TASK_GROUP_SEP: &str = "/";

pub fn make_task_group(domain: &str, intent: &str) -> String {
    format!("{domain}{TASK_GROUP_SEP}{intent}")
}

pub fn parse_task_group(task_group: &str) -> (String, String) {
    let mut parts = task_group.splitn(2, TASK_GROUP_SEP);
    let domain = parts.next().unwrap_or("").to_string();
    let intent = parts.next().unwrap_or("").to_string();
    (domain, intent)
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RoutingObservation {
    pub schema_version: String,
    pub observation_id: String,
    pub arm_id: String,
    pub dispatch_id: String,
    pub task_domain: String,
    pub task_intent: String,
    pub selected_tier: String,
    pub baseline_tier: String,
    pub quality_score: f64,
    pub cost: f64,
    pub latency_ms: i64,
    pub success: bool,
    pub failure_domain: Option<String>,
    pub budget_violation: bool,
    pub observed_at: String,
}

impl RoutingObservation {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RoutingArm {
    pub schema_version: String,
    pub arm_id: String,
    pub experiment_id: String,
    pub tier: String,
    pub profile_id: Option<String>,
    pub traffic_weight: f64,
    pub observations: Vec<RoutingObservation>,
}

impl RoutingArm {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RoutingExperiment {
    pub schema_version: String,
    pub experiment_id: String,
    pub name: String,
    pub task_group: String,
    pub arms: Vec<RoutingArm>,
    pub status: String,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
    pub conclusion: Option<String>,
}

impl RoutingExperiment {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionVerdict {
    pub schema_version: String,
    pub verdict: String,
    pub task_group: String,
    pub candidate_tier: String,
    pub baseline_tier: String,
    pub sample_count: usize,
    pub quality_delta: f64,
    pub cost_reduction_pct: f64,
    pub failure_rate_delta: f64,
    pub reasons: Vec<String>,
    pub requires_human_review: bool,
}

impl PromotionVerdict {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RoutingSelection {
    pub selected_tier: String,
    pub selected_profile_id: Option<String>,
    pub fallback_tier: String,
    pub fallback_profile_id: Option<String>,
    pub shadow_routes: Vec<Value>,
    pub rejected_candidates: Vec<Value>,
    pub routing_reason: String,
    pub routing_mode: String,
    pub routing_experiment_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CostOfPassAggregate {
    pub tier: String,
    pub task_group: String,
    pub total_count: usize,
    pub failure_count: usize,
    pub total_cost: f64,
    pub total_quality: f64,
    pub cost_of_pass: Option<f64>,
}

impl CostOfPassAggregate {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UsageLedgerRow {
    pub row_id: String,
    pub dispatch_id: String,
    pub model_profile_id: String,
    pub cost_of_pass_group: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub estimated_cost: f64,
    pub quality_score: f64,
    pub success: bool,
    pub failure_domain: Option<String>,
    pub latency_ms: i64,
}

impl UsageLedgerRow {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

pub fn parse_cost_of_pass_group(group: &str) -> (String, String, String, String) {
    let parts: Vec<&str> = group.split('/').collect();
    let scope = parts.first().unwrap_or(&"").to_string();
    let family = parts.get(1).unwrap_or(&"").to_string();
    let variant = parts.get(2).unwrap_or(&"").to_string();
    let criterion = parts.get(3).unwrap_or(&"").to_string();
    (scope, family, variant, criterion)
}

pub fn aggregate_cost_of_pass(rows: &[UsageLedgerRow]) -> Option<CostOfPassAggregate> {
    if rows.is_empty() {
        return None;
    }
    let first = &rows[0];
    let tier = first.model_profile_id.clone();
    let group = first.cost_of_pass_group.clone();
    let (_, family, variant, _) = parse_cost_of_pass_group(&group);
    let task_group = make_task_group(&family, &variant);

    let total_count = rows.len();
    let failure_count = rows.iter().filter(|r| !r.success).count();
    let total_cost: f64 = rows.iter().map(|r| r.estimated_cost).sum();
    let total_quality: f64 = rows.iter().map(|r| r.quality_score).sum();

    let cost_of_pass = if total_count > 0 && total_quality > 0.0 {
        Some(total_cost / total_quality)
    } else {
        None
    };

    Some(CostOfPassAggregate {
        tier,
        task_group,
        total_count,
        failure_count,
        total_cost,
        total_quality,
        cost_of_pass,
    })
}
